src/main/java/net/snowflake/ingest/streaming/internal/ParquetRowBuffer.java

```diff
@@ -0,0 +1,595 @@
+
+    PrimitiveType.PrimitiveTypeName type = getTimePrimitiveType(physicalType);
+    if (physicalType == ColumnPhysicalType.SB16) {
+      LogicalTypeAnnotation typeAnnotation = LogicalTypeAnnotation.decimalType(scale, 38);
+      return Types.primitive(type, repetition).as(typeAnnotation).length(16).id(id).named(name);
+    } else {
+      LogicalTypeAnnotation typeAnnotation =
+          logicalType == ColumnLogicalType.TIME
+              ? LogicalTypeAnnotation.timeType(false, timeUnit)
+              : LogicalTypeAnnotation.timestampType(false, timeUnit);
+      return Types.primitive(type, repetition).as(typeAnnotation).id(id).named(name);
+    }
+  }
+
+  /**
+   * Get the parquet primitive type name for column of a Snowflake time logical type.
+   *
+   * @param physicalType Snowflake physical type of column
+   * @return column parquet primitive type name
+   */
+  private static PrimitiveType.PrimitiveTypeName getTimePrimitiveType(
+      ColumnPhysicalType physicalType) {
+    PrimitiveType.PrimitiveTypeName type;
+    switch (physicalType) {
+      case SB4:
+        type = PrimitiveType.PrimitiveTypeName.INT32;
+        break;
+      case SB8:
+        type = PrimitiveType.PrimitiveTypeName.INT64;
+        break;
+      case SB16:
+        type = PrimitiveType.PrimitiveTypeName.FIXED_LEN_BYTE_ARRAY;
+        break;
+      default:
+        throw new UnsupportedOperationException("Time physical type: " + physicalType);
+    }
+    return type;
+  }
+
+  /**
+   * Get the parquet time unit for column of a Snowflake time logical type.
+   *
+   * @param scale Snowflake column scale
+   * @return column parquet time unit
+   */
+  private static LogicalTypeAnnotation.TimeUnit getTimeUnitFromScale(int scale) {
+    LogicalTypeAnnotation.TimeUnit timeUnit;
+    if (scale <= 3) {
+      timeUnit = LogicalTypeAnnotation.TimeUnit.MILLIS;
+    } else if (scale <= 6) {
+      timeUnit = LogicalTypeAnnotation.TimeUnit.MICROS;
+    } else {
+      timeUnit = LogicalTypeAnnotation.TimeUnit.NANOS;
+    }
+    return timeUnit;
+  }
+
+  @Override
+  boolean hasColumn(String name) {
+    return fieldIndex.containsKey(name);
+  }
+
+  @Override
+  float addRow(
+      Map<String, Object> row,
+      int curRowIndex,
+      Map<String, RowBufferStats> statsMap,
+      Set<String> formattedInputColumnNames) {
+    return addRow(row, data, statsMap, formattedInputColumnNames);
+  }
+
+  @Override
+  float addTempRow(
+      Map<String, Object> row,
+      int curRowIndex,
+      Map<String, RowBufferStats> statsMap,
+      Set<String> formattedInputColumnNames) {
+    return addRow(row, tempData, statsMap, formattedInputColumnNames);
+  }
+
+  /**
+   * Adds a row to the parquet buffer.
+   *
+   * @param row row to add
+   * @param out internal buffer to add to
+   * @param statsMap column stats map
+   * @param inputColumnNames list of input column names after formatting
+   * @return row size
+   */
+  private float addRow(
```

Comment:
All this is and the data type specific logic below is tested where :) ?

Reply:
we are working on this in a separate PR to integrate with the test refactoring in #209.
@sfc-gh-gdoci looks into this.

Reply:
I have added tests and merged here.